components/sup/src/manager/service/mod.rs

```diff
@@ -841,9 +859,48 @@ impl Service {
             }
         } else {
             self.check_process();
+
+            // Health Checks are either run on a default or customizable interval (specified at load time).
+            // The check time is stored in `self.scheduled_health_check` as an Option<Instant>
+            // During each 'tick' we compare Instant::now() to scheduled_health_check and run
+            // a new check if scheduled_health_check <= Instant::now() (right now or in the past).
+            // When a health check completes, we record the time in self.last_health_check as an
+            // Option<Instant> and set self.scheduled_health_check to Now + the interval.
+
+            // At any point in the manager loop, we can schedule an 'out of band' health check; for instance,
+            // if we detect that the Service has just been updated or the census has changed. In
+            // those cases `self.schedule_special_health_check()` will effectively change the
+            // scheduled_health_check to Now + _default_ interval. This allows us to provide an updated
+            // health check status for the service given something changed in the ring such as a
+            // depended upon service bouncing, within a reasonable amount of time. When a special check is
+            // scheduled, we set an additional heuristic (self.special_health_check_scheduled) bool to true.
+
+            // In order to avoid perpetually extending the scheduled_health_check in a scenario where
+            // something keeps flapping in the ring, we also check if self.special_health_check_scheduled
+            // is true and the Duration since self.last_health_check exceeds HealthCheckInterval::default()
+            // When that condition is met, we run a health check at that instant.
+
```

Comment:
This logic overall is pretty gnarly. I wonder if we can simplify it by taking a step back. Does this sound correct?

1. We never want to run a health check *more* frequently than `HealthCheckInterval::default()`
2. Unless `self.scheduled_health_check` is set in the future, we want to run it now

If that's true, I think we can simplify this down to:
```rust
            match (self.last_health_check, self.scheduled_health_check) {
                (Some(last_check_instant), _)
                    if last_check_instant.elapsed() < HealthCheckInterval::default() =>
                {
                    trace!(
                        "Skipping health check; it was run only {:?} ago",
                        last_check_instant
                    );
                }
                (_, Some(scheduled_check_instant)) if scheduled_check_instant > now => {
                    trace!(
                        "Skipping health check; next scheduled for {:?} (now: {:?})",
                        scheduled_check_instant,
                        now
                    );
                }
                _ => self.run_health_check_hook(),
            }
```

I think this allows us to do away with the `special_health_check_scheduled` entirely. Am I missing something?

Reply:
Unfortunately I do not think I agree with premise 1. as it feels a bit inflexible with some edge-case users likely requesting more frequent checks in the future, but I can probably bend if that makes the implementation less complex. Reading 2. and thinking..

Reply:
We could go with 1 for now and see if a real use case for more frequent health checks develops in the future. As always, I default to YAGNI when code is getting complex.

Reply:
@baumanj I believe that the following captures all the scenarios and is a good balance between clarity and conciseness. Also, I added some [comments](https://github.com/habitat-sh/habitat/pull/5890/files#diff-320293bb5f68f709add66ac43f5ea52eR863) that I think summarize the implementation as a whole.

```
            match (self.scheduled_health_check, self.last_health_check) {
                (Some(scheduled_check_instant), _) => {
                    // the check time has passed or is now, so we run a check
                    if scheduled_check_instant <= now {
                        self.run_health_check_hook();
                    }
                }
                (_, Some(last_check_instant)) => {
                    // we scheduled a special check so we will honor that contract
                    if self.special_health_check_scheduled
                        && last_check_instant.elapsed() >= *HealthCheckInterval::default().as_ref()
                    {
                        self.run_health_check_hook();
                    }
                }
                _  => self.run_health_check_hook(),
            }
```

Reply:
I'm not really clear why we need the `special_health_check_scheduled` Boolean. It seems to add noticeable complexity, but I'm not seeing the use case that makes it necessary. What functionality do we need _right now_ that's not captured by [these two simple rules](https://github.com/habitat-sh/habitat/pull/5890#discussion_r242289186)?

If we can use that, we have only 3 cases:
1. Don't run the check because it's too frequent
1. Don't run the check because it's not time yet
1. Run the check

With the code you're proposing here there are 5 cases:
1. Run the check because it's at or past the scheduled time
1. Don't run the check because it's before the scheduled time
1. Run the check because there's a special check scheduled and it it's past the default interval since the last check
1. Don't run the check because there's either not a special check scheduled or it's too frequent
1. Run the check

Also, the 3-case version has the benefit of covering all the reasons we *don't* want to run the check first and then letting the catch-all be running the check. The 5-case version has nested conditions which go back and forth between running and not running, which (at least for me) make it harder to reason about.

That's a lot more cognitive overhead, I think. Plus the maintenance of the `special_health_check_scheduled` Boolean through other parts of the code.

Reply:
I believe there is one case that you are not accounting for and is why we need the `special_health_check_scheduled` boolean. 

Since we schedule an 'out of band' health check when something changes in the Supervisor mesh, we need to mitigate a situation where we could perpetually extend the `scheduled_health_check` because the thing is perpetually flapping.  When we schedule an 'out of band' health check we need to guarantee that it will run within `HealthCheckInterval::default()`.

If we mark the fact that a `self.special_health_check_scheduled` is true, then we know that we HAVE to run a new check within `HealthCheckInterval::default()` interval - regardless of what the actual configured interval is.

Reply:
@baumanj I think moving the perpetual re-scheduling mitigation into the `schedule_health_check()` function is prudent. Please take a look at my last commit. I think it makes use of your recommendations. Thank you!

Reply:
It does require [this](https://github.com/habitat-sh/habitat/pull/5890/files#diff-320293bb5f68f709add66ac43f5ea52eR951)

Reply:
I should be able to look at this tomorrow. Thanks for being so open to feedback, as always!